runtime/sema/entitlementset.go

```diff
@@ -0,0 +1,191 @@
+		typeIDs = append(typeIDs, string(entitlementType.ID()))
+	})
+	sort.Strings(typeIDs)
+
+	// Join type IDs
+	var sb strings.Builder
+	for index, typeID := range typeIDs {
+		if index > 0 {
+			sb.WriteByte('|')
+		}
+		sb.WriteString(typeID)
+	}
+	return sb.String()
+}
+
+// DisjunctionOrderedSet is a set of entitlement disjunctions, keyed by disjunctionKey
+type DisjunctionOrderedSet = orderedmap.OrderedMap[string, *EntitlementOrderedSet]
+
+// EntitlementSet is a set (conjunction) of entitlements and entitlement disjunctions.
+// e.g. {entitlements: A, B; disjunctions: (C | D), (E | F)}
+type EntitlementSet struct {
+	// Entitlements is a set of entitlements
+	Entitlements *EntitlementOrderedSet
+	// Disjunctions is a set of entitlement disjunctions, keyed by disjunctionKey
+	Disjunctions *DisjunctionOrderedSet
+}
+
+// Add adds an entitlement to the set.
+//
+// NOTE: The resulting set is potentially not minimal:
+// If the set contains a disjunction that contains the entitlement,
+// then the disjunction is NOT discarded.
+// Call Minimize to obtain a minimal set.
+func (s *EntitlementSet) Add(entitlementType *EntitlementType) {
+	if s.Entitlements == nil {
+		s.Entitlements = orderedmap.New[EntitlementOrderedSet](1)
+	}
+	s.Entitlements.Set(entitlementType, struct{}{})
+}
+
+// AddDisjunction adds an entitlement disjunction to the set.
+// If the set already contains an entitlement of the given disjunction,
+// then the disjunction is discarded.
+func (s *EntitlementSet) AddDisjunction(disjunction *EntitlementOrderedSet) {
+	// If this set already contains an entitlement of the given disjunction,
+	// there is no need to add the disjunction.
+	if s.Entitlements != nil &&
+		disjunction.ForAnyKey(s.Entitlements.Contains) {
+
+		return
+	}
+
+	// If the disjunction already exists in the set,
+	// there is no need to add the disjunction.
+	key := disjunctionKey(disjunction)
+	if s.Disjunctions != nil && s.Disjunctions.Contains(key) {
+		return
+	}
+
+	if s.Disjunctions == nil {
+		s.Disjunctions = orderedmap.New[DisjunctionOrderedSet](1)
+	}
+	s.Disjunctions.Set(key, disjunction)
+}
+
+// Merge merges the other entitlement set into this set.
+// The result is the union of the entitlements and disjunctions of both sets.
+//
+// The result is not necessarily minimal:
+// For example, if s contains a disjunction d,
+// and other contains an entitlement e that is part of d,
+// then the result will still contain d.
+// See Add.
+// Call Minimize to obtain a minimal set.
+func (s *EntitlementSet) Merge(other *EntitlementSet) {
+	if other.Entitlements != nil {
+		other.Entitlements.Foreach(func(key *EntitlementType, _ struct{}) {
+			s.Add(key)
+		})
+	}
+
+	if other.Disjunctions != nil {
+		other.Disjunctions.
+			Foreach(func(_ string, disjunction *EntitlementOrderedSet) {
+				s.AddDisjunction(disjunction)
+			})
+	}
+}
+
+// Minimize minimizes the entitlement set.
+// It removes disjunctions that contain entitlements
```

Comment:
How much extra work would it be to remove "redundant" disjunctions as well? I.e., if we have an `EntitlementSet` that contains `(A | B)` and `(A | B | C)`, the minimal representation of this set would be `(A | B)`.

Reply:
I mainly avoided that (and also minimizing on all operations except the final access conversion) for performance reasons, as e.g. such a minimization would grow with the number of entitlements, disjunctions, and entitlements in the disjunctions. We can maybe add it for "reasonable simple" cases.